Cython/Distutils/build_ext.py

```diff
@@ -10,34 +11,126 @@
+        ('cython-directives=', None,
+            "compiler directive overrides"),
+        ('cython-gdb', None,
+             "generate debug information for cygdb"),
+        ('cython-compile-time-env', None,
+            "cython compile time environment"),
         ])
 
     boolean_options.extend([
-        'cython-c-in-temp'
+        'cython-cplus', 'cython-create-listing', 'cython-line-directives',
+        'cython-c-in-temp', 'cython-gdb',
     ])
 
     def initialize_options(self):
         _build_ext.initialize_options(self)
+        self.cython_cplus = 0
+        self.cython_create_listing = 0
+        self.cython_line_directives = 0
+        self.cython_include_dirs = None
+        self.cython_directives = None
         self.cython_c_in_temp = 0
+        self.cython_gen_pxi = 0
+        self.cython_gdb = False
+        self.cython_compile_time_env = None
+
+    def finalize_options(self):
+        _build_ext.finalize_options(self)
+        if self.cython_include_dirs is None:
+            self.cython_include_dirs = []
+        elif isinstance(self.cython_include_dirs, basestring):
+            self.cython_include_dirs = \
+                self.cython_include_dirs.split(os.pathsep)
+        if self.cython_directives is None:
+            self.cython_directives = {}
 
     def build_extension(self, ext):
         from Cython.Build.Dependencies import cythonize
-        if self.cython_c_in_temp:
+
+        # Set up the include_path for the Cython compiler:
+        #    1.    Start with the command line option.
+        #    2.    Add in any (unique) paths from the extension
+        #        cython_include_dirs (if Cython.Distutils.extension is used).
+        #    3.    Add in any (unique) paths from the extension include_dirs
+        includes = list(self.cython_include_dirs)
+        try:
+            for i in ext.cython_include_dirs:
+                if i not in includes:
+                    includes.append(i)
+        except AttributeError:
+            pass
+
+        # In case extension.include_dirs is a generator, evaluate it and keep
+        # result
+        ext.include_dirs = list(ext.include_dirs)
+        for i in ext.include_dirs:
+            if i not in includes:
+                includes.append(i)
+
+        # Set up Cython compiler directives:
+        #    1. Start with the command line option.
+        #    2. Add in any (unique) entries from the extension
+        #         cython_directives (if Cython.Distutils.extension is used).
+        directives = dict(self.cython_directives)
+        if hasattr(ext, "cython_directives"):
+            directives.update(ext.cython_directives)
+
+        if self.cython_c_in_temp or getattr(ext, 'cython_c_in_temp', 0):
             build_dir = self.build_temp
         else:
             build_dir = None
-        new_ext = cythonize(ext,force=self.force, quiet=self.verbose == 0, build_dir=build_dir)[0]
+
+        if self.cython_cplus or getattr(ext, 'cython_cplus', 0):
```

Comment:
This construct occurs so often that it seems to merit a helper function, e.g. `get_extension_attr(options, extension, option_name, default=False)`, where `options` would probably just be the `build_ext` instance.

Reply:
I have added `get_extension_attr(self, extension, option_name, default=False)` method. I hope now it is better.